File: C#/Pathfinding/Pathfinding.cs
using System;
using System.Collections.Generic;

// Full repository can be found on : https://github.com/Ellandq/Pathfinding/tree/main/C%23/Pathfinding

namespace Pathfinding
{
    public class Pathfinding<T> where T : PathNode
    {
        // Movement Grid
        protected T[,] nodeGrid;

        // Movement Values
        protected const int MOVE_STRAIGHT_COST = 10;
        protected const int MOVE_DIAGONAL_COST = 14;

        public static Pathfinding<T> Instance { get; protected set; }

        public int CheckedNodeCounter { get; protected set; }

        public virtual List<T?> FindPath(int startPosX, int startPosY, int endPosX, int endPosY)
        {
            Console.WriteLine("Pathfinding algorithm not chosen.");
            return null;
        }

        protected void InitializeNodes(T[,] nodeGrid)
        {
            this.nodeGrid = nodeGrid;

            for (var x = 0; x < nodeGrid.GetLength(0); x++)
            {
                for (var y = 0; y < nodeGrid.GetLength(1); y++)
                {
                    this.nodeGrid[x, y].Initialize(x, y);
                }
            }
        }

        protected virtual HashSet<T> GetNeighbourList(T currentNode)
        {
            var neighbourList = new HashSet<T>();

            if (currentNode.GridPosX - 1 >= 0)
            {
                // Left
                neighbourList.Add(nodeGrid[currentNode.GridPosX - 1, currentNode.GridPosY]);
                // Left Down
                if (currentNode.GridPosY - 1 >= 0) neighbourList.Add(nodeGrid[currentNode.GridPosX - 1, currentNode.GridPosY - 1]);
                // Left Up
                if (currentNode.GridPosY + 1 < nodeGrid.GetLength(1)) neighbourList.Add(nodeGrid[currentNode.GridPosX - 1, currentNode.GridPosY + 1]);
            }
            if (currentNode.GridPosX + 1 < nodeGrid.GetLength(0))
            {
                // Right
                neighbourList.Add(nodeGrid[currentNode.GridPosX + 1, currentNode.GridPosY]);
                // Right Down
                if (currentNode.GridPosY - 1 >= 0) neighbourList.Add(nodeGrid[currentNode.GridPosX + 1, currentNode.GridPosY - 1]);
                // Right Up
                if (currentNode.GridPosY + 1 < nodeGrid.GetLength(1)) neighbourList.Add(nodeGrid[currentNode.GridPosX + 1, currentNode.GridPosY + 1]);
            }
            if (currentNode.GridPosX - 1 >= 0)
            {
                // Down
                if (currentNode.GridPosY - 1 >= 0) neighbourList.Add(nodeGrid[currentNode.GridPosX, currentNode.GridPosY - 1]);
                // Up
                if (currentNode.GridPosY + 1 < nodeGrid.GetLength(1)) neighbourList.Add(nodeGrid[currentNode.GridPosX, currentNode.GridPosY + 1]);
            }

            return neighbourList;
        }

        protected List<T?> CalculatePath(T? endNode)
        {
            var path = new List<T?> { endNode };

            var currentNode = endNode;

            while (currentNode.cameFromNode != null)
            {
                path.Add(currentNode.cameFromNode as T);
                currentNode = currentNode.cameFromNode as T;
            }

            path.Reverse();
            
            return path;
        }

        // Calculates a rough estimate of the distance between two positions 
        protected int CalculateDistanceCost(Tuple<int, int> a, Tuple<int, int> b)
        {
            var xDistance = Math.Abs(a.Item1 - b.Item1);
            var yDistance = Math.Abs(a.Item2 - b.Item2);
            var remaining = Math.Abs(xDistance - yDistance);
            return MOVE_DIAGONAL_COST * Math.Min(xDistance, yDistance) + MOVE_STRAIGHT_COST * remaining;
        }

        protected void ReinitalizeGrid(HashSet<T> used)
        {
            foreach (var node in used)
            {
                node.Initialize();
            }
        }
    }
}

File: Java/Pathfinding-Algorithms/src/Pathfinding/Pathfinding.java
package Pathfinding;

import java.util.ArrayList;
import java.util.List;
import Utils.Pair;

public class Pathfinding<T extends PathNode> {

    // Movement Grid
    protected T[][] nodeGrid;

    // Movement Values
    protected final int MOVE_STRAIGHT_COST = 10;
    protected final int MOVE_DIAGONAL_COST = 14;

    public static Pathfinding<PathNode> Instance; // Java does not have static generic types, so using PathNode for the example

    public int CheckedNodeCounter;

    public List<T> findPath(int startPosX, int startPosY, int endPosX, int endPosY) {
        System.out.println("Pathfinding algorithm not chosen.");
        return null;
    }

    protected void initializeNodes(T[][] nodeGrid) {
        this.nodeGrid = nodeGrid;

        for (int x = 0; x < nodeGrid.length; x++) {
            for (int y = 0; y < nodeGrid[x].length; y++) {
                this.nodeGrid[x][y].initialize(x, y);
            }
        }
    }

    protected List<T> getNeighbourList(T currentNode) {
        List<T> neighbourList = new ArrayList<>();

        if (currentNode.getGridPosX() - 1 >= 0) {
            // Left
            neighbourList.add(nodeGrid[currentNode.getGridPosX() - 1][currentNode.getGridPosY()]);
            // Left Down
            if (currentNode.getGridPosY() - 1 >= 0) neighbourList.add(nodeGrid[currentNode.getGridPosX() - 1][currentNode.getGridPosY() - 1]);
            // Left Up
            if (currentNode.getGridPosY() + 1 < nodeGrid[0].length) neighbourList.add(nodeGrid[currentNode.getGridPosX() - 1][currentNode.getGridPosY() + 1]);
        }
        if (currentNode.getGridPosX() + 1 < nodeGrid.length) {
            // Right
            neighbourList.add(nodeGrid[currentNode.getGridPosX() + 1][currentNode.getGridPosY()]);
            // Right Down
            if (currentNode.getGridPosY() - 1 >= 0) neighbourList.add(nodeGrid[currentNode.getGridPosX() + 1][currentNode.getGridPosY() - 1]);
            // Right Up
            if (currentNode.getGridPosY() + 1 < nodeGrid[0].length) neighbourList.add(nodeGrid[currentNode.getGridPosX() + 1][currentNode.getGridPosY() + 1]);
        }
        if (currentNode.getGridPosX() - 1 >= 0) {
            // Down
            if (currentNode.getGridPosY() - 1 >= 0) neighbourList.add(nodeGrid[currentNode.getGridPosX()][currentNode.getGridPosY() - 1]);
            // Up
            if (currentNode.getGridPosY() + 1 < nodeGrid[0].length) neighbourList.add(nodeGrid[currentNode.getGridPosX()][currentNode.getGridPosY() + 1]);
        }

        return neighbourList;
    }

    protected List<T> calculatePath(T endNode) {
        List<T> path = new ArrayList<>();

        path.add(endNode);
        T currentNode = endNode;

        while (currentNode.cameFromNode != null) {
            path.add((T) currentNode.cameFromNode);
            currentNode = (T) currentNode.cameFromNode;
        }

        java.util.Collections.reverse(path);
        return path;
    }

    protected int calculateDistanceCost(Pair<Integer, Integer> a, Pair<Integer, Integer> b) {
        int xDistance = Math.abs(a.value1 - b.value1);
        int yDistance = Math.abs(a.value2 - b.value2);
        int remaining = Math.abs(xDistance - yDistance);
        return MOVE_DIAGONAL_COST * Math.min(xDistance, yDistance) + MOVE_STRAIGHT_COST * remaining;
    }

    protected void reinitalizeGrid(List<T> used) {
        for (T node : used) {
            node.initialize();
        }
    }
}

